### PR TITLE
Allow User-Defined Batching Strategies

### PR DESCRIPTION
**Pull request by dyastremsky**
This PR aims to create tests for Triton's new custom batching feature. The test will use two example custom batching strategies provided in the backend examples folder. The test will start up the server with two instances of the model, then send 12 requests to it. It will check to see if the correct batching strategy was used.

The feature allows the user to pass in the custom batching feature via the model config. If not done so, it will check for a file called "batchstrategy.so" in the model version, model, and backend directories (in that order).

The test checks for the following cases:

1. Single batching is correctly loaded via the backend directory and succeeds.
2. Volume batching is correctly loaded via the backend directory and succeeds.
3. Volume batching is correctly loaded via the model directory and succeeds (with single batching ignored in the backend directory).
4. Volume batching is correctly loaded via the model version directory and succeeds (with single batching ignored in the backend directory).
5. Volume batching is correctly loaded via the model config and succeeds (with single batching ignored in the backend directory).

Core PR: https://github.com/triton-inference-server/core/pull/143/files
Backend PR: https://github.com/triton-inference-server/backend/pull/75

### CODE REVIEW

File: docs/user_guide/model_configuration.md [hunk 1]
```diff
@@ -930,6 +930,14 @@ allow the queue to be configured so that individual requests are
 rejected or deferred if their time in the queue exceeds a specified
 timeout.
 
+#### Custom Batching
+
+You can set custom batching rules that work _in addition to_ the default behavior of the dynamic batcher. To do so, you would implement three functions in [tritonbackend.h](https://github.com/triton-inference-server/core/blob/main/include/triton/core/tritonbackend.h) and create a shared library. These functions are `TRITONBACKEND_ModelBatchIncludeRequest`, `TRITONBACKEND_ModelBatchInitialize`, and `TRITONBACKEND_ModelBatchFinalize`.
```
Comment (nv-kmcgill53) [inline docs/user_guide/model_configuration.md:935]: NIT: I believe we have a max width of 100 chars per line with the exception of links. Please fix.

Comment (dyastremsky) [reply docs/user_guide/model_configuration.md]: Thanks! I didn't know what the hard rule was, and I don't believe our formatter works on .md files. Good to know. Fixed. :)

File: qa/L0_batch_custom/batch_custom_test.py [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+from collections.abc import Iterable
+import tritonclient.grpc as grpcclient
+
+# By default, find tritonserver on "localhost", but can be overridden
+# with TRITONSERVER_IPADDR envvar
+_tritonserver_ipaddr = os.environ.get('TRITONSERVER_IPADDR', 'localhost')
+
+TEST_SYSTEM_SHARED_MEMORY = bool(
+    int(os.environ.get('TEST_SYSTEM_SHARED_MEMORY', 0)))
+TEST_CUDA_SHARED_MEMORY = bool(int(os.environ.get('TEST_CUDA_SHARED_MEMORY',
```
Comment (GuanLuo) [inline qa/L0_batch_custom/batch_custom_test.py:48]: Don't need the handling for shared memory / client variants, only keep the part to test the new functionality

**Review by nv-kmcgill53 — APPROVED** [review]

File: qa/L0_batch_custom/batch_custom_test.py (hunk 2 above)
Comment (dyastremsky) [reply qa/L0_batch_custom/batch_custom_test.py]: Thank you for the feedback. Removed the variants.

**Review by GuanLuo — APPROVED** [review]

**Pull request merged**